app/src/store/loginAndRegister.ts

```diff
@@ -0,0 +1,80 @@
+import { createSlice, type SerializedError } from '@reduxjs/toolkit';
+import { match } from 'ts-pattern';
+import { getAccessToken } from './voice/accessToken';
+import { register } from './voice/registration';
+import { type AsyncStoreSlice } from './app';
+import { createTypedAsyncThunk } from './common';
+import { login, logout } from './user';
+
+export type LoginAndRegisterRejectValue =
+  | {
+      reason: 'LOGIN_REJECTED';
+    }
+  | {
+      reason: 'GET_ACCESS_TOKEN_REJECTED';
+    }
+  | {
+      reason: 'REGISTER_REJECTED';
+    };
+
+export const loginAndRegister = createTypedAsyncThunk<
+  void,
+  void,
+  {
+    rejectValue: LoginAndRegisterRejectValue;
+  }
+>('registration/loginAndRegister', async (_, { dispatch, rejectWithValue }) => {
+  const loginActionResult = await dispatch(login());
+  if (login.rejected.match(loginActionResult)) {
+    return rejectWithValue({ reason: 'LOGIN_REJECTED' });
+  }
+
+  const getAccessTokenResult = await dispatch(getAccessToken());
+  if (getAccessToken.rejected.match(getAccessTokenResult)) {
+    await dispatch(logout());
+    return rejectWithValue({
+      reason: 'GET_ACCESS_TOKEN_REJECTED',
+    });
+  }
+
+  const registerActionResult = await dispatch(register());
+  if (register.rejected.match(registerActionResult)) {
+    return rejectWithValue({ reason: 'REGISTER_REJECTED' });
+  }
+});
+
+export type LoginAndRegisterSlice = AsyncStoreSlice<
+  {},
+  LoginAndRegisterRejectValue | { error: SerializedError }
```

Comment:
I think we'll see the same "poor ergonomics" with the fail state of this slice. Let's add the `reason: UNEXPECTED...` to the default fail state.